pkg/speaker/manager.go

```diff
@@ -0,0 +1,415 @@
+		return m.delLoadBalancer(svc, localRecord)
+	}
+
+	// update service from TypeLoadBalancer to other type
+	if exist && svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		if localRecord == nil {
+			m.Info(fmt.Sprintf("local:%s record exists but no data. changed svc'type is %s", svcNSName, svc.Spec.Type))
+			return nil
+		}
+		m.V(3).Info(fmt.Sprintf("change service:[%s] type from LoadBalancer to %s, so delete LoadBalancer", svc.GetName(), svc.Spec.Type))
+		return m.delLoadBalancer(svc, localRecord)
+	}
+
+	// get service record info
+	svcRecord := m.getSvcRecordInfo(svc)
+	if svcRecord == nil {
+		m.V(3).Info(fmt.Sprintf("get service:[%s] record info error", svc.GetName()))
+		return nil
+	}
+
+	return m.handleSvcBalance(svc, localRecord, svcRecord)
+}
+
+func (m *Manager) handleSvcBalance(svc *corev1.Service, localRecord, svcRecord *recordInfo) error {
+	switch m.getHandleResult(localRecord, svcRecord) {
+	case wantDelete:
+		m.V(1).Info("delLoadBalancer " + localRecord.String())
+		return m.delLoadBalancer(svc, localRecord)
+	case wantStore:
+		m.V(1).Info("setLoadBalancer " + svcRecord.String())
+		return m.setLoadBalancer(svc, svcRecord)
+	case wantReset:
+		m.V(1).Info(fmt.Sprintf("resetLoadBalancer localRecord:%s svcRecord:%s", localRecord.String(), svcRecord.String()))
+		return m.resetLoadBalancer(svc, localRecord, svcRecord)
+
+	default:
+		// endpoint update
+		if svc.Spec.ExternalTrafficPolicy == corev1.ServiceExternalTrafficPolicyTypeLocal {
+			m.V(1).Info("Local svc.externalTrafficPolicy updateLoadBalancer")
+			return m.updateLoadBalancer(svc, svcRecord)
+		}
+		m.V(1).Info("handler do nothing")
+	}
+
+	return nil
+}
+
+func (m *Manager) getHandleResult(localRecord, svcRecord *recordInfo) handleResult {
+	if reflect.DeepEqual(localRecord, svcRecord) {
+		return none
+	}
+
+	if localRecord == nil {
+		return wantStore
+	}
+
+	if svcRecord == nil {
+		return wantDelete
+	}
+
+	m.V(4).Info(fmt.Sprintf("local record is: %s  service record is: %s", localRecord.String(), svcRecord.String()))
+	if !reflect.DeepEqual(localRecord.ips, svcRecord.ips) {
+		if len(svcRecord.ips) == 0 {
+			return wantDelete
+		}
+
+		if len(localRecord.ips) == 0 {
+			return wantStore
+		}
+
+		return wantReset
+	}
+
+	if localRecord.eip != svcRecord.eip {
+		if svcRecord.eip == "" {
+			return wantDelete
+		}
+
+		if localRecord.eip == "" {
+			return wantStore
+		}
+
+		return wantReset
+	}
+
+	if svcRecord.protocols != localRecord.protocols {
+		if svcRecord.protocols == "" {
+			return wantDelete
+		}
+
+		if localRecord.protocols == "" {
+			return wantStore
+		}
+
+		return wantReset
+	}
+
+	return none
+}
+
+func (m *Manager) getSvcRecordInfo(svc *corev1.Service) *recordInfo {
+	if svc == nil || svc.Annotations == nil {
+		return nil
+	}
+
+	r := &recordInfo{}
+	if value, ok := svc.Annotations[constant.OpenELBAnnotationKey]; !ok || value != constant.OpenELBAnnotationValue {
+		return r
```

Comment:
```suggestion
		return nil
```

Reply:
If it returns nil, it means there is something wrong with the passed parameters.
If `&recordInfo{}` is returned, it means that the service has no problem, and then it will judge how to deal with it according to the value of each field. If the fields in `&recordInfo{}` are all "", it means that the user deleted `lb.kubesphere.io/v1alpha1:openelb`.